qadence/ml_tools/constructors.py

```diff
@@ -762,13 +791,28 @@ def build_qnn_from_configs(
         ansatz,
     )
 
+    if observable_config is None:
+        raise ValueError(
+            "You need to provide at least one observable config in the QNN constructor"
+        )
+
     if isinstance(observable_config, list):
         observable = [
             observable_from_config(register=register, config=cfg) for cfg in observable_config
         ]
     else:
         observable = observable_from_config(register=register, config=observable_config)  # type: ignore[assignment]
 
-    ufa = QNN(circ, observable, inputs=fm_config.inputs)
+    ufa = QNN(
+        circ,
+        observable,
+        inputs=inputs,
+        backend=backend,
+        diff_mode=diff_mode,
+        measurement=measurement,
+        noise=noise,
+        configuration=configuration,
+        input_diff_mode=input_diff_mode,
+    )
 
     return ufa
```

Comment:
Sorry for hijacking but I think this function is becoming a bit hard to read. I propose to change the whole body to:

```python
    if observable_config is None:
        raise ValueError(
            "You need to provide at least one observable config in the QNN constructor"
        )

    if ansatz_config is None:
        ansatz_config = AnsatzConfig(depth=0)

    inputs: list[Basic | str] | None = None
    full_fm = chain()
    if fm_config:
        fm_blocks = create_fm_blocks(register=register, config=fm_config)
        full_fm = _interleave_ansatz_in_fm(
            register=register,
            fm_blocks=fm_blocks,
            ansatz_config=ansatz_config,
        )
        inputs = fm_config.inputs

    ansatz = create_ansatz(register=register, config=ansatz_config)

    if ansatz_config.ansatz_strategy == "rydberg":
        # Add a block before the Feature Map to move from 0 state to an
        # equal superposition of all states. This needs to be here only for rydberg
        # feature map and only as long as the feature map is not updated to include
        # a driving term in the Hamiltonian.

        num_qubits = register if isinstance(register, int) else register.n_qubits
        mixing_block = kron(*[H(i) for i in range(num_qubits)])
        full_fm = chain(mixing_block, full_fm)

    circ = QuantumCircuit(register, full_fm, ansatz)

    observable: AbstractBlock | list[AbstractBlock] = (
        [observable_from_config(register=register, config=cfg) for cfg in observable_config]
        if isinstance(observable_config, list)
        else observable_from_config(register=register, config=observable_config)
    )

    return QNN(
        circ,
        observable,
        inputs=inputs,
        backend=backend,
        diff_mode=diff_mode,
        measurement=measurement,
        noise=noise,
        configuration=configuration,
        input_diff_mode=input_diff_mode,
    )
```

Reply:
Although I don't quite understand why fm_config and ansatz_config and observable_config should be allowed to be None. I think we should just not allow for `None`, and then the above simplifies even further

Reply:
```python
    fm_blocks = create_fm_blocks(register=register, config=fm_config)
    full_fm = _interleave_ansatz_in_fm(
        register=register,
        fm_blocks=fm_blocks,
        ansatz_config=ansatz_config,
    )
    ansatz = create_ansatz(register=register, config=ansatz_config)

    if ansatz_config.ansatz_strategy == "rydberg":
        # Add a block before the Feature Map to move from 0 state to an
        # equal superposition of all states. This needs to be here only for rydberg
        # feature map and only as long as the feature map is not updated to include
        # a driving term in the Hamiltonian.

        num_qubits = register if isinstance(register, int) else register.n_qubits
        mixing_block = kron(*[H(i) for i in range(num_qubits)])
        full_fm = chain(mixing_block, full_fm)

    circ = QuantumCircuit(register, full_fm, ansatz)

    observable: AbstractBlock | list[AbstractBlock] = (
        [observable_from_config(register=register, config=cfg) for cfg in observable_config]
        if isinstance(observable_config, list)
        else observable_from_config(register=register, config=observable_config)
    )

    return QNN(
        circ,
        observable,
        inputs=fm_config.inputs,
        backend=backend,
        diff_mode=diff_mode,
        measurement=measurement,
        noise=noise,
        configuration=configuration,
        input_diff_mode=input_diff_mode,
    )
```

Reply:
@awennersteen 
> I think this function is becoming a bit hard to read.
Agreed 100%. The review is much appreciated. Many ad-hoc changes were made to this function.

> Although I don't quite understand why fm_config and ansatz_config and observable_config should be allowed to be None.
It wasn't allowed to be None before today. It is the latest addition. Users using it for Quantum Extremal Learning or other such cases, would want to be able to define a QNN without the feature encoding block.

An alternative is to require the user to specify that the number of features being encoded is 0 and then not have any of these fm_blocks. But with the None option, the user can just not provide a config for the feature map if they don't want any feature map.

Another option, is to make that the default be a feature map with 0 features (and thus have an empty block). The current default is 1 feature. But having a default feature map, which doesn't even do anything could be a bit confusing.

Thoughts?

Reply:
I had somewhat the same though as @awennersteen and commented about the `observable_config`. I think that if it makes the logic simpler, we should avoid optional args that are unnecessarily checked for existence except for the `FeatureMapConfig` if there is a user need as you pointed out. Then, it needs to be moved after the defaulted arguments which might be confusing. In this case, I'd be inclined to default it to 0 features, check for it and skip adding an empty block.